File: sandbox/contribs/ibm_core/java-src/nio/src/com/ibm/io/nio/LockManager.java
/* Copyright 2004 The Apache Software Foundation or its licensors, as applicable
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ibm.io.nio;


import java.nio.channels.FileLock;
import java.nio.channels.OverlappingFileLockException;
import java.util.Comparator;
import java.util.Iterator;
import java.util.SortedSet;
import java.util.TreeSet;

/**
 * The lock manager is responsible for tracking acquired and pending locks on
 * the underlying file channel.
 * 
 */
final class LockManager {
	// The set of acquired and pending locks.
	private final Comparator lockComparator = new Comparator() {
		public int compare(Object lock1, Object lock2) {
			long position1 = ((FileLock) lock1).position();
			long position2 = ((FileLock) lock2).position();
			return position1 > position2 ? 1 : (position1 < position2 ? -1 : 0);
		}
	};

	private final SortedSet locks = new TreeSet(lockComparator);

	/*
	 * Default Constructor.
	 */
	protected LockManager() {
		super();
	}

	/*
	 * Add a new pending lock to the manager. Thows an exception if the lock
	 * would overlap an existing lock. Once the lock is acquired it remains in
	 * this set as an acquired lock.
	 */
	synchronized void addLock(FileLock lock)
			throws OverlappingFileLockException {
		long lockEnd = lock.position() + lock.size();
		for (Iterator keyItr = locks.iterator(); keyItr.hasNext();) {
			FileLock existingLock = (FileLock) keyItr.next();
			if (existingLock.position() > lockEnd) {
				// This, and all remaining locks, start beyond our end (so
				// cannot overlap).
				break;
			}
			if (existingLock.overlaps(lock.position(), lock.size())) {
				throw new OverlappingFileLockException();
			}
		}
		locks.add(lock);
	}

	/*
	 * Removes an acquired lock from the lock manager. If the lock did not exist
	 * in the lock manager the operation is a no-op.
	 */
	synchronized void removeLock(FileLock lock) {
		locks.remove(lock);
	}
}

File: drlvm/modules/vm/src/main/native/vmcore/shared/thread/lock_manager.cpp
/*
 *  Licensed to the Apache Software Foundation (ASF) under one or more
 *  contributor license agreements.  See the NOTICE file distributed with
 *  this work for additional information regarding copyright ownership.
 *  The ASF licenses this file to You under the Apache License, Version 2.0
 *  (the "License"); you may not use this file except in compliance with
 *  the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
#include "port_mutex.h"
#include "lock_manager.h"
#include "vm_threads.h"
#include "exceptions.h"

Lock_Manager::Lock_Manager()
{
    UNREF IDATA stat = port_mutex_create (&lock, APR_THREAD_MUTEX_NESTED);
    assert(stat==TM_ERROR_NONE);
}

Lock_Manager::~Lock_Manager()
{
    UNREF IDATA stat = port_mutex_destroy (&lock);
    assert(stat==TM_ERROR_NONE);
}

void Lock_Manager::_lock()
{
    UNREF IDATA stat = port_mutex_lock(&lock);
    assert(stat==TM_ERROR_NONE);
}

bool Lock_Manager::_tryLock()
{     
    IDATA stat = port_mutex_trylock(&lock);
    return stat==TM_ERROR_NONE;    
}

void Lock_Manager::_unlock()
{
    UNREF IDATA stat = port_mutex_unlock(&lock);
    assert(stat==TM_ERROR_NONE);
}


void Lock_Manager::_lock_enum()
{   
    Lock_Manager::_lock_enum_or_null(0);
}


void Lock_Manager::_unlock_enum()
{
   Lock_Manager::_unlock();
}


//
// If a lock is immediately available return true
//      otherwise return NULL
//
//  Use unlock_or_null so that the names match.
//

bool Lock_Manager::_lock_or_null()
{
    return Lock_Manager::_tryLock();
}


void Lock_Manager::_unlock_or_null()
{
    Lock_Manager::_unlock_enum();
}


void Lock_Manager::_unlock_enum_or_null() 
{
     Lock_Manager::_unlock_enum();
} 

bool Lock_Manager::_lock_enum_or_null(bool UNREF return_null_on_fail)
{
    IDATA stat = port_mutex_lock(&lock);
    return stat==TM_ERROR_NONE;
}
